File: 1046-max-consecutive-ones-iii/1046-max-consecutive-ones-iii.py
from typing import List

class Solution:
    def longestOnes(self, nums: List[int], k: int) -> int:
        max_len, zero, l = 0, 0, 0

        for r in range(len(nums)):
            if nums[r] == 0:
                zero += 1
            
            while zero > k:
                if nums[l] == 0:
                    zero -= 1
                l += 1
            
            max_len = max(max_len, r - l + 1)

        return max_len

File: 1046-max-consecutive-ones-iii/1046-max-consecutive-ones-iii.cpp
class Solution {
public:
    int longestOnes(vector<int>& nums, int k) {
        int maxLen = 0;
        int cntZero = 0;
        int l = 0, r = 0;

        while (r < nums.size()) {
            if (nums[r] == 0)
                cntZero++;

            while (cntZero > k)
                if (nums[l++] == 0)
                    cntZero--;

            if (cntZero <= k)
                maxLen = max(maxLen, r - l + 1);

            r++;
        }

        return maxLen;
    }
};